File: packages/crawler/src/parsers/excel_parser.py
"""Excel解析器"""

import re
from typing import Any, Dict, List, Optional

import pandas as pd
from loguru import logger


class ExcelParser:
    """Excel文件解析器

    功能：
    - 使用pandas读取Excel文件
    - 多Sheet处理
    - 表头识别
    - 合并单元格处理
    - 数据类型转换
    """

    # 标准字段映射（同HTML解析器）
    FIELD_MAPPING = {
        "招录机关": "department_name",
        "部门名称": "department_name",
        "用人单位": "department_name",
        "职位名称": "position_name",
        "岗位名称": "position_name",
        "职位代码": "position_code",
        "岗位代码": "position_code",
        "招录人数": "recruit_count",
        "招考人数": "recruit_count",
        "学历": "education_min",
        "学历要求": "education_min",
        "专业": "major_specific",
        "专业要求": "major_specific",
        "政治面貌": "political_status",
        "年龄": "age_requirement",
        "工作经历": "work_exp_requirement",
        "基层工作经历": "grassroots_exp_years",
        "户籍": "hukou_requirement",
        "性别": "gender_required",
        "工作地点": "work_location",
        "备注": "notes",
        "其他条件": "other_requirements",
    }

    def __init__(self):
        pass

    def parse(
        self,
        file_path: str,
        sheet_name: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """解析Excel文件

        Args:
            file_path: 文件路径
            sheet_name: 指定工作表名称，None表示自动识别

        Returns:
            解析后的数据列表
        """
        try:
            # 读取所有工作表
            excel_file = pd.ExcelFile(file_path)
            sheet_names = excel_file.sheet_names

            logger.info(f"Excel文件包含 {len(sheet_names)} 个工作表: {sheet_names}")

            all_results = []

            for name in sheet_names:
                if sheet_name and name != sheet_name:
                    continue

                df = pd.read_excel(file_path, sheet_name=name, header=None)

                # 检查是否为职位表
                if not self._is_position_sheet(df):
                    logger.debug(f"工作表 '{name}' 不是职位表，跳过")
                    continue

                # 解析工作表
                results = self._parse_sheet(df)

                if results:
                    logger.info(f"工作表 '{name}' 解析出 {len(results)} 条记录")
                    all_results.extend(results)

            return all_results

        except Exception as e:
            logger.error(f"Excel文件解析失败: {file_path} -> {e}")
            raise

    def _is_position_sheet(self, df: pd.DataFrame) -> bool:
        """判断是否为职位表工作表"""
        if df.empty:
            return False

        # 检查前几行是否包含职位表特征
        position_keywords = ["职位", "岗位", "部门", "学历", "专业", "人数", "招录"]

        for row_idx in range(min(5, len(df))):
            row_text = " ".join(str(v) for v in df.iloc[row_idx].values if pd.notna(v))
            match_count = sum(1 for kw in position_keywords if kw in row_text)
            if match_count >= 3:
                return True

        return False

    def _parse_sheet(self, df: pd.DataFrame) -> List[Dict[str, Any]]:
        """解析工作表"""
        # 查找表头行
        header_row = self._find_header_row(df)

        if header_row is None:
            logger.warning("未能找到表头行")
            return []

        # 获取表头
        headers = df.iloc[header_row].values.tolist()

        # 映射字段名
        mapped_headers = self._map_headers(headers)

        # 解析数据行
        results = []

        for row_idx in range(header_row + 1, len(df)):
            row = df.iloc[row_idx]

            # 跳过空行
            if row.isna().all():
                continue

            item = {}
            for col_idx, value in enumerate(row.values):
                if col_idx < len(mapped_headers):
                    field_name = mapped_headers[col_idx]
                    if field_name and pd.notna(value):
                        item[field_name] = self._convert_value(value)

            if item:
                results.append(item)

        return results

    def _find_header_row(self, df: pd.DataFrame) -> Optional[int]:
        """查找表头行"""
        header_keywords = list(self.FIELD_MAPPING.keys())

        for row_idx in range(min(10, len(df))):
            row = df.iloc[row_idx]
            row_values = [str(v) if pd.notna(v) else "" for v in row.values]

            match_count = sum(
                1 for cell in row_values if any(kw in cell for kw in header_keywords)
            )

            if match_count >= 3:
                return row_idx

        return None

    def _map_headers(self, headers: List) -> List[Optional[str]]:
        """映射表头到标准字段名"""
        mapped = []

        for header in headers:
            header_str = str(header).strip() if pd.notna(header) else ""

            if not header_str:
                mapped.append(None)
                continue

            # 精确匹配
            if header_str in self.FIELD_MAPPING:
                mapped.append(self.FIELD_MAPPING[header_str])
                continue

            # 模糊匹配
            matched = False
            for key, value in self.FIELD_MAPPING.items():
                if key in header_str or header_str in key:
                    mapped.append(value)
                    matched = True
                    break

            if not matched:
                field_name = re.sub(r"\s+", "_", header_str.lower())
                mapped.append(field_name)

        return mapped

    def _convert_value(self, value: Any) -> Any:
        """转换数据值"""
        if pd.isna(value):
            return None

        # 数值类型
        if isinstance(value, (int, float)):
            if value == int(value):
                return int(value)
            return value

        # 字符串类型
        value_str = str(value).strip()

        # 清洗字符串
        value_str = re.sub(r"\s+", " ", value_str)

        return value_str

    def get_sheet_names(self, file_path: str) -> List[str]:
        """获取所有工作表名称"""
        excel_file = pd.ExcelFile(file_path)
        return excel_file.sheet_names

    def preview(self, file_path: str, rows: int = 10) -> Dict[str, pd.DataFrame]:
        """预览Excel文件内容"""
        excel_file = pd.ExcelFile(file_path)
        preview_data = {}

        for sheet_name in excel_file.sheet_names:
            df = pd.read_excel(file_path, sheet_name=sheet_name, nrows=rows)
            preview_data[sheet_name] = df

        return preview_data

File: apps/server/internal/parser/excel_parser.go
package parser

import (
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/extrame/xls"
	"github.com/xuri/excelize/v2"
	"go.uber.org/zap"
)

// isHTMLFile checks if a file is actually HTML content (common on gov sites)
func isHTMLFile(filePath string) bool {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return false
	}
	
	// Check for common HTML signatures
	content := strings.ToLower(string(data[:min(1000, len(data))]))
	return strings.Contains(content, "<html") ||
		strings.Contains(content, "<!doctype html") ||
		strings.Contains(content, "<table") ||
		strings.Contains(content, "<head")
}

// extractTextFromHTML extracts text from HTML files disguised as Excel
func (p *ExcelPositionParser) extractTextFromHTML(filePath string) (string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", err
	}
	
	doc, err := goquery.NewDocumentFromReader(bytes.NewReader(data))
	if err != nil {
		return "", err
	}
	
	var builder strings.Builder
	
	// Extract text from tables
	doc.Find("table").Each(func(idx int, table *goquery.Selection) {
		builder.WriteString(fmt.Sprintf("=== Table %d ===\n", idx+1))
		table.Find("tr").Each(func(_ int, row *goquery.Selection) {
			var cells []string
			row.Find("th, td").Each(func(_ int, cell *goquery.Selection) {
				text := strings.TrimSpace(cell.Text())
				cells = append(cells, text)
			})
			if len(cells) > 0 {
				builder.WriteString(strings.Join(cells, "\t"))
				builder.WriteString("\n")
			}
		})
		builder.WriteString("\n")
	})
	
	// If no tables found, extract all text
	if builder.Len() == 0 {
		builder.WriteString(doc.Find("body").Text())
	}
	
	return builder.String(), nil
}

// ExcelPositionParser parses Excel files to extract position data
type ExcelPositionParser struct {
	Logger *zap.Logger
}

// NewExcelParser creates a new Excel parser
func NewExcelParser(logger *zap.Logger) *ExcelPositionParser {
	return &ExcelPositionParser{Logger: logger}
}

// Parse parses an Excel file and extracts positions
// Supports both .xlsx and .xls formats, also handles HTML files disguised as Excel
func (p *ExcelPositionParser) Parse(filePath string) ([]ParsedPosition, error) {
	// First, check if the file is actually HTML disguised as Excel
	if isHTMLFile(filePath) {
		if p.Logger != nil {
			p.Logger.Debug("Detected HTML content in Excel file, using HTML parser",
				zap.String("file", filePath),
			)
		}
		return p.parseHTML(filePath)
	}
	
	ext := strings.ToLower(filepath.Ext(filePath))
	
	// Try appropriate parser based on extension
	if ext == ".xls" {
		positions, err := p.parseXLS(filePath)
		if err == nil && len(positions) > 0 {
			return positions, nil
		}
		// Fallback to HTML parser
		htmlPositions, htmlErr := p.parseHTML(filePath)
		if htmlErr == nil && len(htmlPositions) > 0 {
			return htmlPositions, nil
		}
		if err != nil {
			return nil, fmt.Errorf("failed to parse xls file: %w", err)
		}
		return positions, nil
	}
	
	// For .xlsx files, try excelize first
	positions, err := p.parseXLSX(filePath)
	if err != nil {
		// If xlsx parsing fails, the file might actually be an xls file
		// (common on Chinese government websites that save HTML as .xls/.xlsx)
		if p.Logger != nil {
			p.Logger.Debug("XLSX parsing failed, trying fallback parsers",
				zap.String("file", filePath),
				zap.Error(err),
			)
		}
		
		// Try xls parser as fallback
		xlsPositions, xlsErr := p.parseXLS(filePath)
		if xlsErr == nil && len(xlsPositions) > 0 {
			return xlsPositions, nil
		}
		
		// Try HTML parser as last resort
		htmlPositions, htmlErr := p.parseHTML(filePath)
		if htmlErr == nil && len(htmlPositions) > 0 {
			if p.Logger != nil {
				p.Logger.Debug("File is actually HTML, parsed using HTML parser",
					zap.String("file", filePath),
				)
			}
			return htmlPositions, nil
		}
		
		// All parsers failed, return original error
		return nil, fmt.Errorf("failed to open excel file: %w", err)
	}
	
	return positions, nil
}

// parseHTML parses HTML files disguised as Excel and extracts positions
func (p *ExcelPositionParser) parseHTML(filePath string) ([]ParsedPosition, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}
	
	doc, err := goquery.NewDocumentFromReader(bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	
	var allPositions []ParsedPosition
	
	// Process all tables
	doc.Find("table").Each(func(_ int, table *goquery.Selection) {
		positions := p.parseHTMLTable(table)
		allPositions = append(allPositions, positions...)
	})
	
	return allPositions, nil
}

// parseHTMLTable parses a single HTML table
func (p *ExcelPositionParser) parseHTMLTable(table *goquery.Selection) []ParsedPosition {
	var rows [][]string
	
	// Extract all rows
	table.Find("tr").Each(func(_ int, row *goquery.Selection) {
		var cells []string
		row.Find("th, td").Each(func(_ int, cell *goquery.Selection) {
			cells = append(cells, strings.TrimSpace(cell.Text()))
		})
		if len(cells) > 0 {
			rows = append(rows, cells)
		}
	})
	
	if len(rows) < 2 {
		return nil
	}
	
	// Find header row
	headerRowIdx := -1
	for i, row := range rows {
		if p.isHeaderRow(row) {
			headerRowIdx = i
			break
		}
	}
	
	if headerRowIdx < 0 {
		return nil
	}
	
	// Map headers
	headers := rows[headerRowIdx]
	mappedHeaders := p.mapHeaders(headers)
	
	// Parse data rows
	var positions []ParsedPosition
	for i := headerRowIdx + 1; i < len(rows); i++ {
		row := rows[i]
		if len(row) == 0 {
			continue
		}
		
		position := p.parseRow(row, mappedHeaders)
		if position != nil && position.PositionName != "" {
			positions = append(positions, *position)
		}
	}
	
	return positions
}

// parseXLSX parses .xlsx files using excelize
func (p *ExcelPositionParser) parseXLSX(filePath string) ([]ParsedPosition, error) {
	f, err := excelize.OpenFile(filePath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var allPositions []ParsedPosition

	// Process all sheets
	for _, sheetName := range f.GetSheetList() {
		positions, err := p.parseSheet(f, sheetName)
		if err != nil {
			if p.Logger != nil {
				p.Logger.Warn("Failed to parse sheet",
					zap.String("sheet", sheetName),
					zap.Error(err),
				)
			}
			continue
		}
		allPositions = append(allPositions, positions...)
	}

	if p.Logger != nil {
		p.Logger.Debug("Excel parsing completed",
			zap.String("file", filePath),
			zap.Int("sheets", len(f.GetSheetList())),
			zap.Int("positions", len(allPositions)),
		)
	}

	return allPositions, nil
}

// parseXLS parses .xls files using extrame/xls library
func (p *ExcelPositionParser) parseXLS(filePath string) ([]ParsedPosition, error) {
	xlsFile, err := xls.Open(filePath, "utf-8")
	if err != nil {
		return nil, err
	}

	var allPositions []ParsedPosition

	// Process all sheets
	for sheetIdx := 0; sheetIdx < xlsFile.NumSheets(); sheetIdx++ {
		sheet := xlsFile.GetSheet(sheetIdx)
		if sheet == nil {
			continue
		}

		positions, err := p.parseXLSSheet(sheet)
		if err != nil {
			if p.Logger != nil {
				p.Logger.Warn("Failed to parse XLS sheet",
					zap.String("sheet", sheet.Name),
					zap.Error(err),
				)
			}
			continue
		}
		allPositions = append(allPositions, positions...)
	}

	if p.Logger != nil {
		p.Logger.Debug("XLS parsing completed",
			zap.String("file", filePath),
			zap.Int("positions", len(allPositions)),
		)
	}

	return allPositions, nil
}

// parseXLSSheet parses a single sheet from an XLS file
func (p *ExcelPositionParser) parseXLSSheet(sheet *xls.WorkSheet) ([]ParsedPosition, error) {
	maxRow := int(sheet.MaxRow)
	if maxRow < 2 {
		return nil, nil // Not enough rows
	}

	// Convert sheet to rows format
	var rows [][]string
	for rowIdx := 0; rowIdx <= maxRow; rowIdx++ {
		row := sheet.Row(rowIdx)
		if row == nil {
			rows = append(rows, []string{})
			continue
		}

		var cells []string
		lastCol := row.LastCol()
		for colIdx := 0; colIdx < lastCol; colIdx++ {
			cells = append(cells, row.Col(colIdx))
		}
		rows = append(rows, cells)
	}

	// Find header row
	headerRowIdx := -1
	for i, row := range rows {
		if p.isHeaderRow(row) {
			headerRowIdx = i
			break
		}
	}

	if headerRowIdx < 0 {
		return nil, nil // No valid header found
	}

	// Map headers
	headers := rows[headerRowIdx]
	mappedHeaders := p.mapHeaders(headers)

	// Parse data rows
	var positions []ParsedPosition
	for i := headerRowIdx + 1; i < len(rows); i++ {
		row := rows[i]
		if len(row) == 0 {
			continue
		}

		position := p.parseRow(row, mappedHeaders)
		if position != nil && position.PositionName != "" {
			positions = append(positions, *position)
		}
	}

	return positions, nil
}

// parseSheet parses a single sheet
func (p *ExcelPositionParser) parseSheet(f *excelize.File, sheetName string) ([]ParsedPosition, error) {
	rows, err := f.GetRows(sheetName)
	if err != nil {
		return nil, err
	}

	if len(rows) < 2 {
		return nil, nil // Not enough rows
	}

	// Find header row (first row with position-related keywords)
	headerRowIdx := -1
	for i, row := range rows {
		if p.isHeaderRow(row) {
			headerRowIdx = i
			break
		}
	}

	if headerRowIdx < 0 {
		return nil, nil // No valid header found
	}

	// Map headers
	headers := rows[headerRowIdx]
	mappedHeaders := p.mapHeaders(headers)

	// Parse data rows
	var positions []ParsedPosition
	for i := headerRowIdx + 1; i < len(rows); i++ {
		row := rows[i]
		if len(row) == 0 {
			continue
		}

		position := p.parseRow(row, mappedHeaders)
		if position != nil && position.PositionName != "" {
			positions = append(positions, *position)
		}
	}

	return positions, nil
}

// isHeaderRow checks if a row is a header row
func (p *ExcelPositionParser) isHeaderRow(row []string) bool {
	rowText := strings.Join(row, " ")

	matchCount := 0
	for _, keyword := range PositionTableKeywords {
		if strings.Contains(rowText, keyword) {
			matchCount++
		}
	}

	return matchCount >= 3
}

// mapHeaders maps Excel headers to standard field names
func (p *ExcelPositionParser) mapHeaders(headers []string) []string {
	mapped := make([]string, len(headers))

	for i, header := range headers {
		header = strings.TrimSpace(header)
		if header == "" {
			continue
		}

		// Try exact match
		if fieldName, ok := FieldMapping[header]; ok {
			mapped[i] = fieldName
			continue
		}

		// Try partial match
		for key, fieldName := range FieldMapping {
			if strings.Contains(header, key) || strings.Contains(key, header) {
				mapped[i] = fieldName
				break
			}
		}

		if mapped[i] == "" {
			mapped[i] = normalizeFieldName(header)
		}
	}

	return mapped
}

// parseRow parses a single data row
func (p *ExcelPositionParser) parseRow(row []string, mappedHeaders []string) *ParsedPosition {
	position := &ParsedPosition{
		ParseConfidence: 90,
	}
	hasData := false

	for i, value := range row {
		if i >= len(mappedHeaders) || mappedHeaders[i] == "" {
			continue
		}

		value = strings.TrimSpace(value)
		if value == "" {
			continue
		}

		hasData = true
		p.setPositionField(position, mappedHeaders[i], value)
	}

	if !hasData {
		return nil
	}

	return position
}

// setPositionField sets a field on the position based on field name
func (p *ExcelPositionParser) setPositionField(pos *ParsedPosition, fieldName, value string) {
	switch fieldName {
	case "position_name":
		pos.PositionName = value
	case "department_name":
		pos.DepartmentName = value
	case "department_code":
		pos.DepartmentCode = value
	case "position_code":
		pos.PositionCode = value
	case "recruit_count":
		if count, err := strconv.Atoi(extractNumber(value)); err == nil {
			pos.RecruitCount = count
		}
	case "work_location":
		pos.WorkLocation = value
	case "education_min":
		pos.EducationMin = value
	case "major_specific":
		if strings.Contains(value, "不限") || strings.Contains(value, "无限制") {
			pos.MajorUnlimited = true
		} else {
			pos.MajorSpecific = splitMajors(value)
		}
	case "political_status":
		pos.PoliticalStatus = value
	case "gender_required":
		pos.GenderRequired = normalizeGender(value)
	case "age_requirement":
		pos.AgeRequirement = value
	case "work_exp_requirement":
		pos.WorkExpRequirement = value
	case "hukou_requirement":
		pos.HukouRequirement = value
	case "other_requirements":
		pos.OtherRequirements = value
	case "notes":
		pos.Notes = value
	}
}

// GetSheetNames returns the names of all sheets in the Excel file
func (p *ExcelPositionParser) GetSheetNames(filePath string) ([]string, error) {
	f, err := excelize.OpenFile(filePath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	return f.GetSheetList(), nil
}

// GetSheetData returns raw data from a specific sheet
func (p *ExcelPositionParser) GetSheetData(filePath, sheetName string) ([][]string, error) {
	f, err := excelize.OpenFile(filePath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	return f.GetRows(sheetName)
}

// ExtractText extracts all text content from an Excel file
// Supports both .xlsx and .xls formats, with fallback for misnamed files
// Also handles HTML files disguised as Excel (common on Chinese gov websites)
func (p *ExcelPositionParser) ExtractText(filePath string) (string, error) {
	// First, check if the file is actually HTML disguised as Excel
	if isHTMLFile(filePath) {
		if p.Logger != nil {
			p.Logger.Debug("Detected HTML content in Excel file, using HTML parser",
				zap.String("file", filePath),
			)
		}
		return p.extractTextFromHTML(filePath)
	}
	
	ext := strings.ToLower(filepath.Ext(filePath))
	
	if ext == ".xls" {
		// Try xls parser first
		text, err := p.extractTextFromXLS(filePath)
		if err == nil && text != "" {
			return text, nil
		}
		// Fallback to xlsx parser (some .xls files are actually xlsx)
		xlsxText, xlsxErr := p.extractTextFromXLSX(filePath)
		if xlsxErr == nil && xlsxText != "" {
			return xlsxText, nil
		}
		// Try HTML as last resort
		htmlText, htmlErr := p.extractTextFromHTML(filePath)
		if htmlErr == nil && htmlText != "" {
			return htmlText, nil
		}
		// Return original error
		if err != nil {
			return "", err
		}
		return "", xlsxErr
	}
	
	// For .xlsx files, try excelize first
	text, err := p.extractTextFromXLSX(filePath)
	if err != nil {
		// If xlsx parsing fails, try xls parser as fallback
		if p.Logger != nil {
			p.Logger.Debug("XLSX text extraction failed, trying fallback parsers",
				zap.String("file", filePath),
				zap.Error(err),
			)
		}
		
		xlsText, xlsErr := p.extractTextFromXLS(filePath)
		if xlsErr == nil && xlsText != "" {
			return xlsText, nil
		}
		
		// Try HTML as last resort (some sites save HTML as .xlsx)
		htmlText, htmlErr := p.extractTextFromHTML(filePath)
		if htmlErr == nil && htmlText != "" {
			if p.Logger != nil {
				p.Logger.Debug("File is actually HTML, extracted using HTML parser",
					zap.String("file", filePath),
				)
			}
			return htmlText, nil
		}
		
		// All parsers failed, return original error
		return "", fmt.Errorf("failed to open excel file: %w", err)
	}
	
	return text, nil
}

// extractTextFromXLSX extracts text from .xlsx files
func (p *ExcelPositionParser) extractTextFromXLSX(filePath string) (string, error) {
	f, err := excelize.OpenFile(filePath)
	if err != nil {
		return "", fmt.Errorf("failed to open excel file: %w", err)
	}
	defer f.Close()

	var builder strings.Builder

	// Process all sheets
	for _, sheetName := range f.GetSheetList() {
		rows, err := f.GetRows(sheetName)
		if err != nil {
			continue
		}

		builder.WriteString(fmt.Sprintf("=== Sheet: %s ===\n", sheetName))
		
		for _, row := range rows {
			// Join cells with tab separator
			line := strings.Join(row, "\t")
			if strings.TrimSpace(line) != "" {
				builder.WriteString(line)
				builder.WriteString("\n")
			}
		}
		builder.WriteString("\n")
	}

	return builder.String(), nil
}

// extractTextFromXLS extracts text from .xls files (legacy Excel format)
func (p *ExcelPositionParser) extractTextFromXLS(filePath string) (string, error) {
	xlsFile, err := xls.Open(filePath, "utf-8")
	if err != nil {
		return "", fmt.Errorf("failed to open xls file: %w", err)
	}

	var builder strings.Builder

	// Process all sheets
	for sheetIdx := 0; sheetIdx < xlsFile.NumSheets(); sheetIdx++ {
		sheet := xlsFile.GetSheet(sheetIdx)
		if sheet == nil {
			continue
		}

		builder.WriteString(fmt.Sprintf("=== Sheet: %s ===\n", sheet.Name))

		// Get max row
		maxRow := int(sheet.MaxRow)
		for rowIdx := 0; rowIdx <= maxRow; rowIdx++ {
			row := sheet.Row(rowIdx)
			if row == nil {
				continue
			}

			var cells []string
			lastCol := row.LastCol()
			for colIdx := 0; colIdx < lastCol; colIdx++ {
				cell := row.Col(colIdx)
				cells = append(cells, cell)
			}

			line := strings.Join(cells, "\t")
			if strings.TrimSpace(line) != "" {
				builder.WriteString(line)
				builder.WriteString("\n")
			}
		}
		builder.WriteString("\n")
	}

	return builder.String(), nil
}
